### PR TITLE
Proposal: SQL based AssetIndex / AssetLoader

### LINKED ISSUE #863: SQL: implementation of the `AssetIndex` (+`AssetLoader`)
An SQL-based implementation of the `AssetIndex`/`AssetLoader` is required.

### Requirements
1. Table Schema: `Asset` and `DataAddress` are in separate tables. Having the `id` and the `name` as dedicated columns might be advisable.
2. Indexing: since querying the AssetIndex is very flexible, as a starting point there should be an index on the `id` and `name`
3. Querying: the query implementation must be flexible enough to accomodate a `QuerySpec`.
4. The implementation should implement `AssetIndex` and `AssetLoader`, `accept()` should be ACID.

_Please provide a written proposal for requirement 1 and 3 before beginning the implementation._

- [ ] Design Proposal: #908 
- [ ] Implementation: (PR reference goes here)

### PR DESCRIPTION
**Pull request by denisneuling**
## What this PR changes/adds

This PR adds a technical proposal to realize a SQL based AssetIndex / AssetLoader.

## Linked Issue(s)

Closes #863 

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [x] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

<sub> Denis Neuling <denis.neuling@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

### CODE REVIEW

**Comment by paullatzelsperger** [comment]
just saw, that you already added the soft-delete, which is tracked in a separate issue (#814), which is not scoped for M3. While the effort is laudable, the problem with this is that the other SQL backends and storage technologies (CosmosDB, in-memory) will not have soft-delete for now, so for M3 the `SqlAssetIndex` will behave differently from the rest. 
To avoid that, I think we should skip the soft-delete feature for now.

**Comment by denisneuling** [comment]
@paullatzelsperger Thanks for pointing this out - just saw the concerning interface had been extended and thought it would have been a good idea to already include this, but sure, let's carve out the removal functionality for now. 

As I am somewhat busy with other tasks I am handing over this PR to @bcronin90 - Would you like to continue on this?

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+![ER Diagram](http://www.plantuml.com/plantuml/png/XP513e8m44NtFSMiJNe1GWWhDnwY6Ud4b40XdS58XBjRnI346dVD_DVVtpyb2mOPsaQH5oSZaAqCCfF0NG4Su7KspcR04fo_nA7MQbPVlB4eYDO6Olvn5_ByU2gAnU89zA7hAOWZpa3e9X6ekVCHIJtmfiU_xglnQ3osj8yE7t4PcoismmfjzUvSKoLxVHznBMkQj4vtiO7c22N-eqhoDgxW0G00)
+
+## Table schema
+
+```sql
+
+-- table: edc_asset
+CREATE TABLE IF NOT EXISTS edc_asset (
+    asset_id VARCHAR(255) NOT NULL,
+    deleted BOOLEAN DEFAULT FALSE,
```
Comment (denisneuling) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:12]: @bcronin90 This line also needs to be deleted if soft-delete is now within the scope of this PR.

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+![ER Diagram](http://www.plantuml.com/plantuml/png/XP513e8m44NtFSMiJNe1GWWhDnwY6Ud4b40XdS58XBjRnI346dVD_DVVtpyb2mOPsaQH5oSZaAqCCfF0NG4Su7KspcR04fo_nA7MQbPVlB4eYDO6Olvn5_ByU2gAnU89zA7hAOWZpa3e9X6ekVCHIJtmfiU_xglnQ3osj8yE7t4PcoismmfjzUvSKoLxVHznBMkQj4vtiO7c22N-eqhoDgxW0G00)
+
+## Table schema
+
+```sql
+
+-- table: edc_asset
+CREATE TABLE IF NOT EXISTS edc_asset (
+    asset_id VARCHAR(255) NOT NULL,
```
Comment (paullatzelsperger) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:11]: lets stick with `VARCHAR` and add a comment, that this DDL is only tested with Postgres. There really is no portable way for variable length text columns across multiple RDBMS. For example, `VARCHAR(MAX)` does not even work on Postgres.

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    PRIMARY KEY (asset_id)
+);
+
+
+-- table: edc_asset_dataaddress
+CREATE TABLE IF NOT EXISTS edc_asset_dataaddress (
```
Comment (paullatzelsperger) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:17]:
```suggestion
CREATE TABLE IF NOT EXISTS edc_data_address (
```

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+COMMENT ON COLUMN edc_asset_dataaddress.properties is 'DataAddress properties serialized as JSON';
+
+
+-- table: edc_asset_property
+CREATE TABLE IF NOT EXISTS edc_asset_property (
+    asset_id VARCHAR(255) NOT NULL,
+    property_name VARCHAR(255) NOT NULL,
+    property_string_value VARCHAR(MAX) DEFAULT NULL,
+    property_object_value VARCHAR(MAX) DEFAULT NULL,
```
Comment (paullatzelsperger) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:32]: I guess `property_string_value` and `property_object_value` could be collated into one, given that we'll have a `property_type` field containing the java class name

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+```sql
+BEGIN;
+
+INSERT INTO edc_asset (asset_id) VALUES ("784cdfde-a50c-11ec-b909-0242ac120002");
+INSERT INTO edc_asset_property (asset_id, property_name, property_string_value) VALUES ("784cdfde-a50c-11ec-b909-0242ac120002", "asset:prop:name", "Max Power");
+INSERT INTO edc_asset_property (asset_id, property_name, property_string_value) VALUES ("784cdfde-a50c-11ec-b909-0242ac120002", "asset:prop:description", "Homer to the Max");
+INSERT INTO edc_asset_dataaddress (asset_id, properties) VALUES ("784cdfde-a50c-11ec-b909-0242ac120002", "{ \"uri\": \"https://projects.eclipse.org/proposals/eclipse-dataspace-connector\" }")
+
+INSERT INTO edc_asset (asset_id) VALUES ("80b69de0-a50c-11ec-b909-0242ac120002");
+INSERT INTO edc_asset_property (asset_id, property_name, property_string_value) VALUES ("80b69de0-a50c-11ec-b909-0242ac120002", "asset:prop:name", "Seneca");
```
Comment (paullatzelsperger) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:116]: multiple `VALUES( ...)` expresseions can be chained, see [here](https://www.postgresqltutorial.com/postgresql-insert-multiple-rows)

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        </code>
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+would translate into following DML:
+
+```sql
+BEGIN;
```
Comment (paullatzelsperger) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:108]: We won't have to express transactions explicitly. That what we have the `TransactionContext` (and related classes) for.

File: extensions/sql/asset/sql_asset_index_loader_proposal.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+```
+
+#### AssetLoader: `Asset deleteById(String assetId)`
+Soft deletion has been postponed. For noe, deletion will work as follows.
+
+```sql
+DELETE FROM edc_asset WHERE asset_id = "784cdfde-a50c-11ec-b909-0242ac120002";
+```
+
+
```
Comment (paullatzelsperger) [inline extensions/sql/asset/sql_asset_index_loader_proposal.md:133]: Missing the `findById` case. Note: don't forget to join with data address and properties.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]

**Comment by jimmarino** [comment]
@paullatzelsperger is this PR still relevant or has it been superseded?

**Comment by paullatzelsperger** [comment]
> @paullatzelsperger is this PR still relevant or has it been superseded?

It's already implemented and documented, lets just close this.

**Pull request closed without merging**